File: messages-scala/src/main/scala/org/ybiquitous/messages/scala/MessageKey.scala
package org.ybiquitous.messages.scala

import java.util.Locale

import org.ybiquitous.messages.{ MessageKey => MsgKey }
import org.ybiquitous.messages.{ MessageBuilder => MsgBuilder }

final class MessageKey private(
  val key: String,
  private val resource: Option[String] = None,
  private val builder: Option[MessageBuilder] = None
) {

  private def unwrapArg(arg: Any): AnyRef = arg.asInstanceOf[AnyRef]

  def get(args: Any*) = delegate.get(args map unwrapArg: _*)
  def get(locale: Locale, args: Any*) = delegate.get(locale, args map unwrapArg: _*)

  def getOrElse(defaultValue: String) = delegate.getOrElse(defaultValue)
  def getOrElse(args: Seq[Any], defaultValue: String) = delegate.getOrElse(args map unwrapArg toArray, defaultValue)
  def getOrElse(locale: Locale, defaultValue: String) = delegate.getOrElse(locale, defaultValue)
  def getOrElse(locale: Locale, args: Seq[Any], defaultValue: String) = delegate.getOrElse(locale, args map unwrapArg toArray, defaultValue)

  override def toString = delegate.toString

  private val delegate = resource match{
    case Some(r) => MsgKey.of(key, r)
    case _       => builder match {
      case Some(b) => MsgKey.of(key, new MsgBuilderImpl(b))
      case _       => MsgKey.of(key)
    }
  }

  private class MsgBuilderImpl(builder: MessageBuilder) extends MsgBuilder {
    override def build(locale: Locale, key: String, args: java.lang.Object*): String = {
      builder.build(locale, key, args map wrapArg)
    }

    override def buildOrElse(locale: Locale, key: String, args: Array[java.lang.Object], defaultValue: String): String = {
      builder.buildOrElse(locale, key, args map wrapArg, defaultValue)
    }

    private def wrapArg(arg: java.lang.Object): Any = arg.asInstanceOf[Any]
  }
}

object MessageKey {
  def apply(
    key: String,
    resource: Option[String] = None,
    builder: Option[MessageBuilder] = None
  ): MessageKey = new MessageKey(key, resource, builder)
}

File: messages-core/src/main/java/org/ybiquitous/messages/MessageKey.java
package org.ybiquitous.messages;

import java.util.Locale;
import java.util.Map;
import java.util.WeakHashMap;

public final class MessageKey {

    private static final Object[] EMPTY_ARGS = {};

    private static final MessageBuilder DEFAULT_BUILDER = DefaultMessageBuilder.create(Constants.DEFAULT_MESSAGE_RESOURCE_NAME);

    public static MessageKey of(String key) {
        return of(key, DEFAULT_BUILDER);
    }

    public static MessageKey of(String key, String resourceName) {
        return of(key, DefaultMessageBuilder.create(resourceName));
    }

    public static MessageKey of(String key, MessageBuilder builder) {
        final CacheKey cacheKey = new CacheKey(key, builder);
        MessageKey result = CACHE.get(cacheKey);
        if (result == null) {
            result = new MessageKey(key, builder);
            CACHE.put(cacheKey, result);
        }
        return result;
    }

    private static final Map<CacheKey, MessageKey> CACHE = new WeakHashMap<CacheKey, MessageKey>();

    private final String _key;

    private final MessageBuilder _builder;

    private MessageKey(String key, MessageBuilder builder) {
        this._key = Utils.notNull(key, "key");
        this._builder = Utils.notNull(builder, "builder");
    }

    public String get(Object... args) throws MessageKeyNotFoundException {
        return get(ThreadLocalLocaleHolder.get(), args);
    }

    public String get(Locale locale, Object... args) throws MessageKeyNotFoundException {
        return this._builder.build(locale, this._key, args);
    }

    public String getOrElse(String defaultValue) {
        return getOrElse(EMPTY_ARGS, defaultValue);
    }

    public String getOrElse(Object[] args, String defaultValue) {
        return getOrElse(ThreadLocalLocaleHolder.get(), args, defaultValue);
    }

    public String getOrElse(Locale locale, String defaultValue) {
        return getOrElse(locale, EMPTY_ARGS, defaultValue);
    }

    public String getOrElse(Locale locale, Object[] args, String defaultValue) {
        return this._builder.buildOrElse(locale, this._key, args, defaultValue);
    }

    public String key() {
        return this._key;
    }

    @Override
    public String toString() {
        return this._key + ", " + this._builder;
    }
}
